### DISCUSSION TITLE
Config 相关配置信息是不是可以按需要配置吗？

### DISCUSSION THREAD

**Discussion by houseme** [polarismesh/polaris#396] [Ideas]
What is the feature you want to add?
比如，某个时候只需要注册服务和配置中心功能，这个时候是不是可以设置消费者服务不开启配置？在不需要服务消费服务时，还会生成服务发现的一些文件信息，这样会有点怪怪

Why do you want to add this feature?
实现配置信息，可以按需加载。

@chuntaojun @andrewshan [@Guo-Zhang](https://github.com/Guo-Zhang) @lepdou  @daheige 

issues: https://github.com/polarismesh/polaris-go/issues/61

开发者群讨论：
智刚：
```html
https://github.com/polarismesh/polaris-go/issues/61 @张乐  @春少 @andrewshan  咱们的sdk中的config配置可以按需配置吗？

比如：只需要配置中心和服务注册的配置，是不是服务发现的配置可以不初始化。在仅使用到配置中心的功能时，这个时候还加载了服务发现的功能，比如生成`polaris/backup/`目录信息。这样有点乖乖的。
这个服务配置信息可以拆分吗？
```

春少：
```html
拆分应该是可以
北极星有二次寻址的能力，应该去不掉的
如果不开启二次寻址的话
```

大黑哥：
```html
设计的好的，确实是应该每个组件能按需要加载，启动。每个组件需要从设计上进行解耦合，而不是绑定在一块。尤其在不知道用户怎么使用，以及使用的场景可能不一样，就不能强制绑架用户使用习惯，这点也正是组件可自由组合，可正交化的思维模式。

我比较倾向于组合的思维，让用户选择需要哪些功能，在服务启动时候，要么可以通过WithXxx参数设置交给用户侧来选择用什么，要么提供默认选项，然后用户可修改的原则设计。

组合和正交化设计的思维，在做这些公共组件，框架服务真的很重要，最怕设计者提前为用户包罗了所有部分，这是我最不喜欢的一点。

嗯嗯，可配置化，让用户选择

```

iGuo:
```html
我平时跑Django最头疼的事情就是必须要配置一堆东西
比如数据库必须带
有时候我只是想要一个简单的API，后面再逐步加
一开始调整settings就要花费大量时间
我有两种建议方案
一种是上面的按需启动
一种是给defalut
```

大黑哥：
```html
嗯嗯，可配置化，让用户选择
所以目前北极星确实需要从架构上，思考这些问题，既然开源了，就不能提前把所有的组件部分替用户做了选择，好的设计不仅要考虑易用性，拓展性，还要考虑一下用户实际选择如何使用。
```

小骏：
```html
北极星当前是插件化按需启动的，之前因为有二次寻址功能，所以强制依赖了服务发现模块，这部分功能当前可以关闭，所以理论上如果被调方API的话，是不需要加载的
这个后台功能已经支持，可以做个接口层的拆解
北极星Java其实已经做了这部分拆分了
嗯，配置文件还是要做到模块化，可拆可和
```

大黑哥：
```html
是的，以接口层来处理可以
```

**Comment by Guo-Zhang** [comment]
最理想的配置方案是用户不需要做任何配置。有两个方案：
1. 尽可能通过内置默认配置解决，用户可以默认不带配置文件使用。
2. 给个默认配置模板（Django的方案），让用户可以方便在此基础上修改。

**Reply by houseme** [reply]
> 最理想的配置方案是用户不需要做任何配置。有两个方案：
> 
> 1. 尽可能通过内置默认配置解决，用户可以默认不带配置文件使用。
> 2. 给个默认配置模板（Django的方案），让用户可以方便在此基础上修改。

有默认的配置模版，最好是可以按需加载功能

**Reply by Guo-Zhang** [reply]
无论采取哪种方案，一定要充分地对各种可能的配置方案做最大程度的单元测试覆盖，并且在文档里给读者指路。在社区运营的过程中也要强调收集用户的配置方案，并持续优化。

**Comment by daheige** [comment]
LGTM!


------------------ 原始邮件 ------------------
发件人: Guo Zhang ***@***.***&gt;
发送时间: 2022年5月30日 21:16
收件人: polarismesh/polaris ***@***.***&gt;
抄送: heige ***@***.***&gt;, Mention ***@***.***&gt;
主题: Re: [polarismesh/polaris] Config 相关配置信息是不是可以按需要配置吗？ (Discussion #396)





 
无论采取哪种方案，一定要充分地对各种可能的配置方案做最大程度的单元测试覆盖，并且在文档里给读者指路。在社区运营的过程中也要强调收集用户的配置方案，并持续优化。
